documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,888 @@
+
+![Tunnel learning](./images/dash-tunnel-learning.svg)
+
+To address this problem, the source information of each encap **MUST** be saved in the flow and compared during flow match. These information should no be part of the key, but whenever they change, we should reprocess the packet and update the saved reverse tunnel in the existing flow.
+
+##### 5.6.2.2. Asymmetrical encap handling
+
+There could be cases where the encaps are asymmetrical, which means the incoming packet and return packet uses different encaps. For example, in the case below, the encap of step 2 and 4 are not symmetrical. This means we could never recreate the flow on the DASH pipeline 2 side to tunnel the packet into the extra hop first with our current flow creation logic.
+
+![Asymmetrical encap](./images/dash-asymmetrical-encap.svg)
+
+To fix this issue, a special action called `reverse_tunnel` is defined, which enables reverse side of the encaps.
+
+#### 5.6.3. Flow resimulation
+
+When certain policies are changed, we might need to update the flows that is associated with the policies. For example, when a VM is moved to another place, the VNET mapping entry will be changed, and we need to update the existing flows that is associated with the VNET mapping entry. Otherwise, the outgoing traffic will be tunneled to incorrect place. This is called flow resimulation.
+
+However, the flow resimulation is not as simple as removing the flow and let the packet to going through the pipeline again. For example:
+
+- When a ACL is changed in the outbound side, all flows needs to be resimulated. However, this cannot be implemented by removing all flow entries and let all flows to be recreated with the latest policy. The reason is that ACLs can be asymmetric and flows are created in pair. If the flows are removed, the inbound side traffic will be dropped immediately, because there is no inbound flow to make it bypass the ACLs.
+- Certain actions might need to bypass the flow resimulation to maintain the per flow consistency. For example, in load balancer case, we will want to ensure that flow resimulation will not causing us to forward existing connections to a different backend server.
+
+These requires us to implement the flow resimulation in a more sophisticated way, which is not fully modeled in DASH today. But we will come back to this design later.
+
+### 5.7. Pre-pipeline ACL and Post-pipeline ACL
+
+Pre-pipeline ACL and Post-pipeline ACL are used to drop the unexpected traffic before and after the packet transformation. It works as below:
+
+1. As the high-level pipeline shows above, if an incoming packet hits a flow, it will skip matching all the ACLs. If an incoming packet is denied by a ACL, the packet will be dropped without creating a flow.
```

Comment:
```suggestion
1. As the high-level pipeline shows above, if an incoming packet hits a flow (i.e. is in the FastPath), it will skip matching all the ACLs. If an incoming packet is denied by a ACL, the packet will be dropped without creating a flow.
```